sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala

```diff
@@ -499,8 +499,35 @@ case class TruncateTableCommand(
         val path = new Path(location.get)
         try {
           val fs = path.getFileSystem(hadoopConf)
+          val fileStatus = fs.getFileStatus(path)
+          // Not all fs impl. support these APIs.
+          val optPermission = Try(fileStatus.getPermission())
+          val optAcls = Try(fs.getAclStatus(path).getEntries)
+
           fs.delete(path, true)
+
+          // We should keep original permission/acl of the path.
+          // For owner/group, only super-user can set it, for example on HDFS. Because
+          // current user can delete the path, we assume the user/group is correct or not an issue.
           fs.mkdirs(path)
+          optPermission.foreach { permission =>
+            Try(fs.setPermission(path, permission)) match {
+              case Failure(e) if e.isInstanceOf[UnsupportedOperationException] =>
+                log.warn(s"Can not set original permission $permission back to " +
+                  s"the created path: $path. Exception: ${e.getMessage}")
+              case Failure(e) => throw e
```

Comment:
do we have to fail if we can't set the permission back?

Reply:
In internal discussion about this issue, we think it is more dangerous to have wrong permission/ACLs silently on a secure cluster when truncate table.  This failure tells users what is wrong and they could so something manually to fix it.

Reply:
Again, we can consider to add one config that users can choose to fail if we cannot preserve original permission/ACLs. For non secure cluster, it can be ignored.

Reply:
Makes sense to me, I don't think we need a config. Shall we also fail if the FS can't set permission?

Reply:
We already fail if the FS cannot set permission. No?

Reply:
```
case Failure(e) if e.isInstanceOf[UnsupportedOperationException] =>
                log.warn(s"Can not set original permission $permission back to " +
                  s"the created path: $path. Exception: ${e.getMessage}")
```
it's just a warning

Reply:
Oh, I see. You meant `UnsupportedOperationException` case. Hmm, `setAcl` also has `UnsupportedOperationException` case. I'm not sure if it is possible that we can get permission/ACLs, but `setPermission`/`setAcl` throws `UnsupportedOperationException`.

Logically, following above reason, it should be safer to fail it too if it is `UnsupportedOperationException`.